### PR TITLE
[Serializer] Document the encoders

### PR DESCRIPTION
**Pull request by GuilhemN**
| Q | A |
| --- | --- |
| Doc fix? | no |
| New docs? | yes |
| Applies to | 2.7 -> master |
| Fixed tickets |  |

Extracted from https://github.com/symfony/symfony-docs/pull/6428

### CODE REVIEW

File: components/serializer/introduction.rst [hunk 1]
```diff
@@ -44,7 +44,7 @@ Usage
 
 Using the Serializer component is really simple. You just need to set up
 the :class:`Symfony\\Component\\Serializer\\Serializer` specifying
-which Encoders and Normalizer are going to be available::
+which Encoders and Normalizers are going to be available::
```
Comment (xabbuh) [inline components/serializer/introduction.rst]: I would do the opposite and use "encoders" and "normalizers" here.

File: components/serializer/introduction.rst [hunk 2]
```diff
@@ -58,10 +58,14 @@ which Encoders and Normalizer are going to be available::
 
 The preferred normalizer is the
 :class:`Symfony\\Component\\Serializer\\Normalizer\\ObjectNormalizer`, but other
-normalizers are available.
-To read more about them, refer to the `Normalizers`_ section of this page. All
+normalizers are available. All
 the examples shown below use the ``ObjectNormalizer``.
 
+.. note::
```
Comment (xabbuh) [inline components/serializer/introduction.rst]: could as well be `seealso`

File: components/serializer/introduction.rst [hunk 3]
```diff
@@ -58,10 +58,14 @@ which Encoders and Normalizer are going to be available::
 
 The preferred normalizer is the
 :class:`Symfony\\Component\\Serializer\\Normalizer\\ObjectNormalizer`, but other
-normalizers are available.
-To read more about them, refer to the `Normalizers`_ section of this page. All
+normalizers are available. All
 the examples shown below use the ``ObjectNormalizer``.
```
Comment (xabbuh) [inline components/serializer/introduction.rst]: should be moved to the previous line to fit the 72 characters boundary

File: components/serializer/encoders.rst [hunk 4]
```diff
@@ -0,0 +1,105 @@
+.. index::
+   single: Serializer, Encoders
+
+Encoders
+========
+
+Encoders basically turn **arrays** into **formats** and vice versa.
+They implement :class:`Symfony\\Component\\Serializer\\Encoder\\EncoderInterface` for encoding (array to format) and :class:`Symfony\\Component\\Serializer\\Encoder\\DecoderInterface` for decoding (format to array).
```
Comment (xabbuh) [inline components/serializer/encoders.rst]: Please wrap lines after the first word that crosses the 72nd character (seem below).

File: components/serializer/encoders.rst [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+You can add new encoders to a Serializer instance by using its second constructor argument::
+
+    use Symfony\Component\Serializer\Serializer;
+    use Symfony\Component\Serializer\Encoder\XmlEncoder;
+    use Symfony\Component\Serializer\Encoder\JsonEncoder;
+
+    $encoders = array(new XmlEncoder(), new JsonEncoder());
+    $serializer = new Serializer(array(), $encoders);
+
+Built-in encoders
```
Comment (xabbuh) [inline components/serializer/encoders.rst]: Encoders

File: components/serializer/encoders.rst [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+-----------------
+
+You can see in the example above that we use two encoders:
```
Comment (xabbuh) [inline components/serializer/encoders.rst]: We always avoid the first person perspective.

File: components/serializer/encoders.rst [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+The ``JsonEncoder``
+~~~~~~~~~~~~~~~~~~~
+
+The ``JsonEncoder`` is much simpler and is based on the PHP `json_encode`_ and `json_decode`_ functions.
+
+.. _json_encode: https://secure.php.net/manual/fr/function.json-encode.php
+.. _json_decode: https://secure.php.net/manual/fr/function.json-decode.php
+
+Custom encoders
```
Comment (xabbuh) [inline components/serializer/encoders.rst]: Encoders

File: components/serializer/encoders.rst (hunk 6 above)
Comment (xabbuh) [inline components/serializer/encoders.rst]: same below

File: components/serializer/encoders.rst [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    // will return:
+    // <?xml version="1.0"?>
+    // <response>
+    //     <foo bar="value" />
+    // </response>
+
+The ``JsonEncoder``
+~~~~~~~~~~~~~~~~~~~
+
+The ``JsonEncoder`` is much simpler and is based on the PHP `json_encode`_ and `json_decode`_ functions.
```
Comment (xabbuh) [inline components/serializer/encoders.rst]: You can use the `phpfunction` role instead.

File: components/serializer/encoders.rst [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+.. _json_encode: https://secure.php.net/manual/fr/function.json-encode.php
+.. _json_decode: https://secure.php.net/manual/fr/function.json-decode.php
+
+Custom encoders
+---------------
+
+If you need to support another format than XML and JSON, you can create your own encoder.
+We will guess that you want to serialize and deserialize Yaml. For that, we will use
+:doc:`/components/yaml/index`::
```
Comment (xabbuh) [inline components/serializer/encoders.rst]: Maybe we should move the "Custom Encoders" section to its own cookbook article instead?

**Comment by ogizanagi** [comment]
Should we document the context options used by each of the built-in encoders ?

**Comment by Ener-Getick** [comment]
@xabbuh Thanks for the comments, they should be fixed now.

@ogizanagi I won't have the time to do much more for now but of course it could be a great addition !

**Comment by xabbuh** [comment]
👍 

Status: Reviewed

**Comment by Ener-Getick** [comment]
@xabbuh i forgot your last comment, i'll move the custom encoders section to its own cookbook tonight.
Status: Needs work

**Comment by Ener-Getick** [comment]
I moved the ``Custom encoders`` section to its own cookbook. @xabbuh is it still ok for you ? :)

Status: Needs review

File: cookbook/map.rst.inc [hunk 10]
```diff
@@ -187,9 +187,10 @@
   * :doc:`/cookbook/security/securing_services`
   * :doc:`/cookbook/security/access_control`
 
-* **Serializer**
+* :doc:`/cookbook/serializer/index`
 
-  * :doc:`/cookbook/serializer`
+  * :doc:`/cookbook/serializer/introduction`
```
Comment (Ener-Getick) [inline cookbook/map.rst.inc]: not sure about this name

**Comment by weaverryan** [comment]
Merged via #6976. Thanks @Ener-Getick!

**Comment by Ener-Getick** [comment]
Thanks ! Sorry i didn't take the time to solve the conflicts...

**Pull request merged**